### PR TITLE
Multiplayer support for Atomic

### PR DESCRIPTION
**Pull request by keithjohnston**

### CODE REVIEW

**Comment by JoshEngebretson** [comment]
@AtomicBuildBot test this please

**Comment by JoshEngebretson** [comment]
The Android build is failing, though Linux/Emscripten/OSX/iOS pass, any thoughts?

```
[ 80%] Building CXX object Source/Atomic/CMakeFiles/Atomic.dir/Network/MasterServerClient.cpp.o
/Users/jenkins/Home/workspace/AtomicDS_BuildBinaries_AndroidPlayer/Source/Atomic/Network/MasterServerClient.cpp: In member function 'void Atomic::MasterServerClient::CheckForMessageFromMaster()':
/Users/jenkins/Home/workspace/AtomicDS_BuildBinaries_AndroidPlayer/Source/Atomic/Network/MasterServerClient.cpp:222:66: error: expected ')' before 'SCNd32'
                     sscanf(masterMessageLengthStr.CString(), "%" SCNd32, &bytesRemainingInMasterServerMessage_);
```

**Comment by keithjohnston** [comment]
OK I think I fixed it. Can you see if that builds now?

**Comment by JoshEngebretson** [comment]
Builds on all platforms, awesome!  I've scrolled through the code and it looks really solid.  I am going to go over it a little closer tomorrow for any possible comments.  Great work!

File: Script/Packages/Atomic/Network.json [hunk 1]
```diff
@@ -2,7 +2,5 @@
 	"name" : "Network",
 	"sources" : ["Source/Atomic/Network"],
 	"includes" : ["<Atomic/Network/Protocol.h>", "<Atomic/Scene/Scene.h>"],
-	"classes" : ["Network", "NetworkPriority", "HttpRequest"]
-
-
+	"classes" : ["Network", "NetworkPriority", "HttpRequest", "Connection"]
 }
```
Comment (JoshEngebretson) [inline Script/Packages/Atomic/Network.json]: Nice, adding Connection to scripting...

File: Source/Atomic/Network/Connection.cpp [hunk 2]
```diff
@@ -200,6 +205,8 @@ void Connection::SetScene(Scene* newScene)
         scene_->StopAsyncLoading();
         SubscribeToEvent(scene_, E_ASYNCLOADFINISHED, HANDLER(Connection, HandleAsyncLoadFinished));
     }
+
+    SubscribeToEvent(scene_, E_COMPONENTREMOVED, HANDLER(Connection, HandleComponentRemoved));
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/Connection.cpp]: This seems like something that we may want to pass upstream to Urho3D?  I am surprised we needed to add it, seems core.  Does stock Urho3D leak component states without this?

File: Source/Atomic/Network/Connection.cpp [hunk 3]
```diff
@@ -437,6 +444,9 @@ bool Connection::ProcessMessage(int msgID, MemoryBuffer& msg)
     case MSG_PACKAGEINFO:
         ProcessPackageInfo(msgID, msg);
         break;
+    case MSG_STRING:
+        ProcessStringMessage(msgID, msg);
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/Connection.cpp]: Very nice

File: Source/Atomic/Network/Connection.cpp [hunk 4]
```diff
@@ -1546,4 +1578,42 @@ void Connection::ProcessPackageInfo(int msgID, MemoryBuffer& msg)
...
+int Connection::GetControlDataInt(const String &key) {
+    return controls_.extraData_[key].GetInt();
+}
+
+void Connection::SendStringMessage(const String& message)
+{
+    // Send the identity map now
+    VectorBuffer msg;
+    msg.WriteString(message);
+    SendMessage(MSG_STRING, true, true, msg);
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/Connection.cpp]: Great use of VectorBuffer

File: Source/Atomic/Network/Connection.cpp [hunk 5]
```diff
@@ -1546,4 +1578,42 @@ void Connection::ProcessPackageInfo(int msgID, MemoryBuffer& msg)
...
+}
+
+void Connection::ProcessStringMessage(int msgID, MemoryBuffer &msg) {
+    using namespace NetworkMessage;
+
+    VariantMap &eventData = GetEventDataMap();
+    eventData[P_MESSAGEID] = (int) msgID;
+    eventData[P_CONNECTION] = this;
+    eventData[P_DATA] = msg.ReadString();
+    SendEvent(E_NETWORKSTRINGMESSAGE, eventData);
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/Connection.cpp]: I do something similar with IPC, also good use on GetEventDataMap()

File: Source/Atomic/Network/MasterServerClient.cpp [hunk 6]
```diff
@@ -0,0 +1,556 @@
+//
+// Created by Keith Johnston on 4/16/16.
+//
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/MasterServerClient.cpp]: There needs to be a license block on new source files, for example: https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Source/AtomicTool/AtomicTool.cpp#L1

You can copy that and apply your name also add yourself to the AUTHORS.md :)  https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/AUTHORS.md

File: Source/Atomic/Network/MasterServerClient.cpp [hunk 7]
```diff
@@ -0,0 +1,556 @@
...
+{
+    if (clientIdToPunchThroughSocketMap_.Size() > 0)
+    {
+        if (timeTillNextPunchThroughAttempt_ <= 0)
+        {
+            for (HashMap<String, kNet::Socket*>::Iterator i = clientIdToPunchThroughSocketMap_.Begin();
+                 i != clientIdToPunchThroughSocketMap_.End();)
+            {
+                Network* network = GetSubsystem<Network>();
+                LOGINFO("Sending packet to client");
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/MasterServerClient.cpp]: Can we remove this logging?

File: Source/Atomic/Network/MasterServerClient.cpp [hunk 8]
```diff
@@ -0,0 +1,556 @@
...
+
+    // See if we need to still be trying to punch through from the client
+    if (clientConnectToGameServerState_ == GAME_CONNECTING_EXTERNAL_IP &&
+        clientToServerSocket_ != NULL && timeTillNextClientConnectAttempt_ <= 0)
+    {
+        Atomic::Network* network = GetSubsystem<Network>();
+
+        if (!network->GetServerConnection() || !network->GetServerConnection()->IsConnected())
+        {
+            LOGINFO("Sending packet to server");
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/MasterServerClient.cpp]: Would be good to remove this logging too, there are also log levels like LOGDEBUG and LOGERROR to use

File: Source/Atomic/Network/MasterServerClient.cpp [hunk 9]
```diff
@@ -0,0 +1,556 @@
...
+                    s->Send("K",1);
+                    ++i;
+                }
+            }
+
+            // Reset the timer
+            timeTillNextPunchThroughAttempt_ = timeBetweenClientPunchThroughAttempts_;
+        }
+
+        timeTillNextPunchThroughAttempt_ -= dt;
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/MasterServerClient.cpp]: This punch through stuff boggles me :)

File: Source/Atomic/Network/MasterServerClient.cpp [hunk 10]
```diff
@@ -0,0 +1,556 @@
...
+        while (n < totalBytes) {
+            char c = buf->buffer.buf[n++];
+
+            // Are we still reading in the length?
+            if (readingMasterMessageLength) {
+                if (c == ':') {
+                    sscanf(masterMessageLengthStr.CString(), "%u" , &bytesRemainingInMasterServerMessage_);
+                    readingMasterMessageLength = false;
+
+                    LOGINFO("Message is " + String(bytesRemainingInMasterServerMessage_) + " long");
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/MasterServerClient.cpp]: Logging to remove

File: Source/Atomic/Network/MasterServerClient.cpp [hunk 11]
```diff
@@ -0,0 +1,556 @@
...
+        LOGERROR("Could not connect to master server");
+        SendEvent(E_MASTERCONNECTIONFAILED);
+    }
+
+    if (state == MASTER_CONNECTED)
+    {
+        SendEvent(E_MASTERCONNECTIONREADY);
+    }
+
+    connectToMasterState_ = state;
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/MasterServerClient.cpp]: Great work, I am pleasantly surprised to find a full master server implementation nestled in this PR :+1:

File: Source/Atomic/Network/MasterServerClient.cpp [hunk 12]
```diff
@@ -0,0 +1,556 @@
...
+                                                 masterUDPConnection_->LocalEndPoint(),
+                                                 masterUDPConnection_->LocalAddress(), serverEndPoint, "",
+                                                 kNet::SocketOverUDP, kNet::ClientConnectionLessSocket, 1400);
+
+        Atomic::Network* network = GetSubsystem<Network>();
+        network->ConnectWithExistingSocket(clientToServerSocket_, remoteGameServerInfo_.clientScene);
+
+        connectToGameServerSecondsRemaining_ = 5.0f;
+
+        LOGINFO("Connecting to Game Server on Internal IP: " +
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/MasterServerClient.cpp]: Probably a good use of LOGDEBUG or LOGDEBUGF here

File: Source/Atomic/Network/MasterServerClient.cpp [hunk 13]
```diff
@@ -0,0 +1,556 @@
...
+                                                 masterUDPConnection_->LocalEndPoint(),
+                                                 masterUDPConnection_->LocalAddress(), serverEndPoint, "",
+                                                 kNet::SocketOverUDP, kNet::ClientConnectionLessSocket, 1400);
+
+        Atomic::Network* network = GetSubsystem<Network>();
+        network->ConnectWithExistingSocket(clientToServerSocket_, remoteGameServerInfo_.clientScene);
+
+        connectToGameServerSecondsRemaining_ = 5.0f;
+
+        LOGINFO("Connecting to Game Server on External IP: " +
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/MasterServerClient.cpp]: Another LOGDEBUG candidate

File: Source/Atomic/Network/MasterServerClient.cpp [hunk 14]
```diff
@@ -0,0 +1,556 @@
...
+
+void MasterServerClient::HandleMasterServerMessage(const String &msg)
+{
+    LOGINFO("Got master server message: " + msg);
+
+    rapidjson::Document document;
+    if (document.Parse<0>(msg.CString()).HasParseError())
+    {
+        LOGERROR("Could not parse JSON data from string");
+        return;
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/MasterServerClient.cpp]: Good error checking

File: Source/Atomic/Network/MasterServerClient.h [hunk 15]
```diff
@@ -0,0 +1,128 @@
+//
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/MasterServerClient.h]: License block needed

File: Source/Atomic/Network/MasterServerClient.h [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+/// Game server info used by a client when connecting
+struct RemoteGameServer
+{
+    /// Game server id
+    String serverId;
+    String internalAddress;
+    unsigned short internalPort;
+    String externalAddress;
+    unsigned short externalPort;
+    Scene* clientScene;
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/MasterServerClient.h]: Any chance of this being released while connecting?  If so, should it be a SharedPtr or WeakPtr depending on whether a strong reference is required?

File: Source/Atomic/Network/MasterServerClient.h [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+    void SendMessageToMasterServer(const String& message);
+    void HandleMasterServerMessage(const String& msg);
+
+    void RegisterServerWithMaster(const String& serverName);
+
+    float udpConnectionSecondsRemaining_;
+    float udpSecondsTillRetry_;
+
+    // Communication with Master Server
+    bool readingMasterMessageLength;
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/MasterServerClient.h]: member variables should have a _ post pended, many of the members have one, some are missing

File: Source/Atomic/Network/Network.cpp [hunk 18]
```diff
@@ -203,6 +204,42 @@ void Network::ClientDisconnected(kNet::MessageConnection* connection)
     }
 }
 
+bool Network::ConnectSimple(const String& address, unsigned short port, Scene* scene)
+{
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/Network.cpp]: Nice and simple!

File: Source/Atomic/Network/Network.cpp [hunk 19]
```diff
@@ -203,6 +204,42 @@ void Network::ClientDisconnected(kNet::MessageConnection* connection)
     }
 }
 
+bool Network::ConnectSimple(const String& address, unsigned short port, Scene* scene)
+{
+    return Connect(address, port, scene, Variant::emptyVariantMap);
+}
+
+bool Network::ConnectWithExistingSocket(kNet::Socket* existingSocket, Scene* scene)
+{
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/Network.cpp]: This is nice too

File: Source/Atomic/Network/Network.cpp [hunk 20]
```diff
@@ -250,9 +287,12 @@ bool Network::StartServer(unsigned short port)
 
     PROFILE(StartServer);
 
+    serverPort_ = port;
+
     if (network_->StartServer(port, kNet::SocketOverUDP, this, true) != 0)
     {
         LOGINFO("Started server on port " + String(port));
+
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/Network.cpp]: There is a fair amount of log, maybe go through and remove what was generated while debugging and change the ones to keep to LOGDEBUG variants?

File: Source/Atomic/Network/Network.cpp [hunk 21]
```diff
@@ -596,6 +654,49 @@ void Network::ConfigureNetworkSimulator()
         i->second_->ConfigureNetworkSimulator(simulatedLatency_, simulatedPacketLoss_);
 }
 
+void Network::ClientConnectToMaster(const String& address, unsigned short port)
+{
+    masterServerClient_.ConnectToMaster(address, port);
```
Comment (JoshEngebretson) [inline Source/Atomic/Network/Network.cpp]: I am not sure we should have master server functionality so closely tied to the Network subsystem.  It is possible that a different master server setup would be desired with the Networking system using whatever matchmaking system was in use, for example Steam.  Do you have thoughts on splitting this out?

File: Source/AtomicMasterServer/package.json [hunk 22]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "atomic-master-server",
```
Comment (JoshEngebretson) [inline Source/AtomicMasterServer/package.json]: Wow, an Atomic Master Server NPM package!  AWESOME!

File: Source/ThirdParty/kNet/include/kNet/Network.h [hunk 23]
```diff
@@ -81,6 +86,11 @@ class Network
 		free it by letting the refcount go to 0. */
 	Ptr(MessageConnection) Connect(const char *address, unsigned short port, SocketTransportLayer transport, IMessageHandler *messageHandler, Datagram *connectMessage = 0);
 
+	// BEGIN ATOMIC CHANGE
+	/** Connect with an existing socket. This is used when creating a connection with NAT punchthrough. */
+	Ptr(MessageConnection) Connect(Socket* s, IMessageHandler *messageHandler, Datagram *connectMessage = 0);
+	// END ATOMIC CHANGE
+
```
Comment (JoshEngebretson) [inline Source/ThirdParty/kNet/include/kNet/Network.h]: Nice job modifying KNet with comments for punch through :)

**Comment by JoshEngebretson** [comment]
Great PR!  I have some notes from looking it over.  The main thing is that I think it would be good to factor the master server stuff out of the Network subsystem and make sure it is optional for network connections, thoughts?

File: Source/Atomic/Network/Connection.cpp (hunk 2 above)
Comment (keithjohnston) [inline Source/Atomic/Network/Connection.cpp]: This was to solve the issue with Javascript component removal not happening until they were garbage collected. By subscribing to this event we can immediately clear out the replication state for the components and the removal event will then be propagated across the network.

File: Source/Atomic/Network/MasterServerClient.cpp (hunk 6 above)
Comment (keithjohnston) [inline Source/Atomic/Network/MasterServerClient.cpp]: Whoops!  I will do that.

File: Source/Atomic/Network/MasterServerClient.cpp (hunk 7 above)
Comment (keithjohnston) [inline Source/Atomic/Network/MasterServerClient.cpp]: Yes.

File: Source/Atomic/Network/MasterServerClient.cpp (hunk 8 above)
Comment (keithjohnston) [inline Source/Atomic/Network/MasterServerClient.cpp]: Will do.

File: Source/Atomic/Network/MasterServerClient.cpp (hunk 9 above)
Comment (keithjohnston) [inline Source/Atomic/Network/MasterServerClient.cpp]: Yeah it's kind of a network hack to be honest. You are tricking the routers into thinking that they each started talking to each other first.

File: Source/Atomic/Network/MasterServerClient.h (hunk 16 above)
Comment (keithjohnston) [inline Source/Atomic/Network/MasterServerClient.h]: Good question. Let me ponder that.

File: Source/Atomic/Network/MasterServerClient.h (hunk 17 above)
Comment (keithjohnston) [inline Source/Atomic/Network/MasterServerClient.h]: Gotcha.

File: Source/Atomic/Network/Network.cpp (hunk 18 above)
Comment (keithjohnston) [inline Source/Atomic/Network/Network.cpp]: Once we have Variant maps exposed on the methods (to javascript), we won't need this.

File: Source/Atomic/Network/Network.cpp (hunk 21 above)
Comment (keithjohnston) [inline Source/Atomic/Network/Network.cpp]: Good point. So maybe create a Matchmaking system and move all the MasterClient stuff into that? Call it AtomicMatchMaking?

**Comment by keithjohnston** [comment]
Thanks for the extensive comments! I will work on these and update the PR accordingly!

**Comment by JoshEngebretson** [comment]
@AtomicBuildBot add to whitelist, retest this please

**Comment by JoshEngebretson** [comment]
@AtomicBuildBot retest this please

Hopefully VS15 Update 2 fixes intermittent build issue with msbuild

**Comment by JoshEngebretson** [comment]
Working on this over yonder: https://github.com/AtomicGameEngine/AtomicGameEngine/tree/JME-ATOMIC-MULTIPLAYER

Should have it landed today 🚊  Choo! Choo!

**Pull request closed without merging**